### PR TITLE
Replace step/indexlist with from/to

### PR DESCRIPTION
**Pull request by troibe**
**What?**
Instead of executing a list of steps/indices we now specify where in the flowgraph we start `-from` and up to which point `-to` we execute.

**Why?**
Using the step and indexlist we didn't actually traverse the flowgraph so some flowgraph semantic was not captured before.

**How?**
We traverse the flowgraph to identify all nodes that can be reached and the nodes that successfully executed.

**Testing**
Used existing tests so far but can add more.

### CODE REVIEW

File: siliconcompiler/remote/client.py
Comment (gadfort) [inline siliconcompiler/remote/client.py:177]: I don't think you can assume any ordering here.
You probably need to use exit nodes as the 'to' and the outputs from the entry nodes as a 'from'

File: siliconcompiler/remote/client.py
Comment (gadfort) [inline siliconcompiler/remote/client.py:313]: fundamentally, this needs to be fixed when we re-work the server.

File: siliconcompiler/scheduler.py
Comment (gadfort) [inline siliconcompiler/scheduler.py:115]: why is this needed, if we provide `arg_step` and `arg_index`

File: siliconcompiler/schema/schema_cfg.py
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:2467]: start flowgraph execution from

File: siliconcompiler/schema/schema_cfg.py
Comment (gadfort) [inline siliconcompiler/schema/schema_cfg.py:2485]: clarify that this is inclusive

File: tests/apps/test_sc.py
Comment (gadfort) [inline tests/apps/test_sc.py:14]: you don't need `-from` here

File: tests/core/test_archive.py
Comment (gadfort) [inline tests/core/test_archive.py:88]: why is this removed? this doesn't change

File: tests/core/test_check_flowgraph.py
Comment (gadfort) [inline tests/core/test_check_flowgraph.py:21]: test for then the from and to are not connected is also needed.

File: tests/core/test_summary.py
Comment (gadfort) [inline tests/core/test_summary.py:26]: this test should be corrected (it looks to have been wrong before too), just picking the previous does not guarantee its the previous node.

File: tests/flows/test_flowstatus.py
Comment (gadfort) [inline tests/flows/test_flowstatus.py:21]: why is this one removed?

File: tests/flows/test_flowstatus.py
Comment (gadfort) [inline tests/flows/test_flowstatus.py:68]: you could just make the list above to and from so it's clearer

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:2532]: this should not change, if step and or index is provided, only those should be archived.

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:3073]: is sorting needed?

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:4006]: remove node_list use if it's not needed

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:4153]: should be formatted as `cts0, cts1, and cts2`

**Review by gadfort — CHANGES_REQUESTED** [review]

File: tests/flows/test_flowstatus.py
Comment (gadfort) [reply tests/flows/test_flowstatus.py:21]: okay, can you add a unit test to ensure a forked run will be marked a failure?

File: siliconcompiler/remote/client.py
Comment (troibe) [reply siliconcompiler/remote/client.py:313]: Honestly I don't know if this actually needed.
There was some logic like that too in siliconcompiler itself for the step/indexlist...

File: siliconcompiler/core.py
Comment (troibe) [reply siliconcompiler/core.py:3073]: Removed it from the description

File: tests/flows/test_flowstatus.py
Comment (troibe) [reply tests/flows/test_flowstatus.py:21]: `test_long_branch` tests the same disjoint branches with failures and I changed the behavior there.
I think this would be duplicate.

File: siliconcompiler/remote/client.py
Comment (troibe) [reply siliconcompiler/remote/client.py:177]: Changed it so it starts off at any successor of an entry node.

File: siliconcompiler/remote/client.py
Comment (troibe) [reply siliconcompiler/remote/client.py:313]: I'll try if I can get it to work without.

File: siliconcompiler/scheduler.py
Comment (troibe) [reply siliconcompiler/scheduler.py:115]: Sure I can just make it such that `arg_step` overwrites `from` and `to` it's value.

**Comment by troibe** [comment]
So I looked into it but at least if we allow multiple disjoint flowgraphs within the same flowgraph,
then this will only catch it if `from`/`to` are explicitely defined.
By default the two single nodes `import` and `syn` without a connection are both `entry` and `exit` nodes.
Therefore the default implied `from=['import', 'syn']` and `to=['import', 'syn']`.
I don't think this is probably not what a novice users intends so we should really reconsider whether we want to allow multiple disjoint flowgraphs within the same flowgraph,

If we explicitely set `from=['import']` and `to=['syn']`. It will catch as expected.

File: siliconcompiler/core.py
Comment (troibe) [inline siliconcompiler/core.py:3857]: This is the preserving of `step/index` args that I was just talking about in our standup.

File: siliconcompiler/remote/client.py
Comment (troibe) [inline siliconcompiler/remote/client.py:171]: Same thing here.

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:1861]: can this be `f'{node[0]}{node[1]}'`

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:1862]: Maybe update to say: `These final nodes in {flow} can not be reached`? I'll leave that up to you

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:2576]: steplist? maybe update the function signature to be "steps="?

**Review by gadfort — COMMENTED** [review]
I think this looks good. Please check the documentation to see if we discuss steplist or indexlist there?

**Comment by troibe** [comment]
@gadfort The last commit removes the remaining mentions of steplist and makes it more clear if all `flowgraph_nodes` are used or just the `nodes_to_execute`.
Please proofread the docstring for `run()`.
I think we should be pretty much ready now.

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:3998]: can you make a ticket to track the remoteid thing? I'm not 100% sure this is correct, but I can't think of how it is wrong.

File: siliconcompiler/core.py
Comment (gadfort) [inline siliconcompiler/core.py:4158]: steps -> nodes

**Review by gadfort — APPROVED** [review]
LGTM, just correct the docstring for `run()`

**Comment by gadfort** [comment]
@troibe please do not merge this as this needs to go along with the #1896

File: siliconcompiler/core.py
Comment (nmoroze) [inline siliconcompiler/core.py:4249]: steps -> nodes?

File: siliconcompiler/report/summary_table.py
Comment (nmoroze) [inline siliconcompiler/report/summary_table.py:72]: I think you need to incorporate index here, e.g.

```python
for val, step, index in chip.schema._getvals(...):
     if not step or (step, index) in flowgraph_nodes:
```

File: siliconcompiler/scheduler.py
Comment (nmoroze) [inline siliconcompiler/scheduler.py:119]: Seems like unrelated change. Why is this being removed?

File: tests/apps/test_sc.py
Comment (nmoroze) [inline tests/apps/test_sc.py:14]: Should this be `-to import` instead? (won't break test either way, but should make it faster)

File: tests/flows/test_from_to.py
Comment (nmoroze) [inline tests/flows/test_from_to.py:40]: typo: afrom->from

**Review by nmoroze — CHANGES_REQUESTED** [review]
Found a couple subtle things that I think be fixed, but overall this looks great, thanks!

Another high-level note is I noticed a lot of casting to `set()` where I think it may not be needed. I think it would be cleaner to remove, but it does seem safer and not dangerous from a functionality perspective to keep those... maybe someday in the future if we add type annotations it will be more comfortable to get rid of those :)

File: siliconcompiler/scheduler.py
Comment (troibe) [reply siliconcompiler/scheduler.py:119]: That was to make debugging easier and is now merged: https://github.com/siliconcompiler/siliconcompiler/pull/1950

File: siliconcompiler/core.py
Comment (troibe) [reply siliconcompiler/core.py:3029]: Sure I can do that. Maybe I should that logic back in if `summary()` is considered to be part of the python API?
Currently that can be done with `chip.set('option', 'from', ['step'])` so removing it also works fine.

File: siliconcompiler/core.py
Comment (troibe) [reply siliconcompiler/core.py:4062]: Thanks for pointing it out. The idea is to operate on the top level dictionary so I'll make it consistent with that

File: siliconcompiler/core.py
Comment (troibe) [reply siliconcompiler/core.py:4158]: You mean 'from' nodes and 'to' nodes?
So far those options are step based so writing nodes here might be confusing to the user.

File: siliconcompiler/apps/sc_remote.py
Comment (troibe) [reply siliconcompiler/apps/sc_remote.py:139]: Yes that should be handled correctly. Internally [when checking the flowgraph entry nodes](https://github.com/siliconcompiler/siliconcompiler/blob/from-to-rebase/siliconcompiler/core.py/#L4729-L4730) we just get the nodes that are [in these from steps](https://github.com/siliconcompiler/siliconcompiler/blob/from-to-rebase/siliconcompiler/core.py/#L4714-L4716). So if there are duplicates in the from steps it doesn't make a difference. (Maybe it would be a tiny bit slower if we had huge graphs)

File: siliconcompiler/core.py
Comment (gadfort) [reply siliconcompiler/core.py:4158]: good point

File: siliconcompiler/core.py
Comment (troibe) [reply siliconcompiler/core.py:4001]: SC-636

File: siliconcompiler/core.py
Comment (nmoroze) [reply siliconcompiler/core.py]: `summary()` is considered part of the API, but I'm overall happy with simplifying the API if we don't see a good use case for this (easier to add back in later than to delete). 

The one use-case I recall of the `steplist` arg was in `sc.py`, but I can't remember why we used it there when I'm pretty sure just reading from the schema is effectively the same functionality wise.

File: siliconcompiler/report/summary_table.py
Comment (nmoroze) [inline siliconcompiler/report/summary_table.py:72]: Watch out, also need to add in `index` to the for loop (otherwise this is using a stale value of index I believe)

**Review by nmoroze — CHANGES_REQUESTED** [review]
Thanks for addressing the feedback! Just one more thing.

File: siliconcompiler/report/summary_table.py
Comment (troibe) [reply siliconcompiler/report/summary_table.py]: Thanks for catching that one.

**Review by nmoroze — APPROVED** [review]

**Pull request merged**
